### PR TITLE
Properly dump directories to tar

### PR DESCRIPTION
**Pull request by MichaelEischer**
What does this PR change? What problem does it solve?
-----------------------------------------------------
The `dump` command previously wrote directories in a tar file in a way which can cause compatibility problems. It accidentally forced the usage of the GNU header format for directories. 7zip on Windows fails to read such directories and in addition it is not possible to dump directories with extended attributes.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
#2319

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- ~~[ ] I have added documentation for the changes (in the manual)~~
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by DRON-666** [comment]
Maybe this bug for another PR but you mentioned 7z compatibility.
tar produced by restic and containing non ASCII names can't be decoded by 7z:
```
>dir /s /b tmp 
\tmp\Folder 2
\tmp\Папка 1
\tmp\Folder 2\File 2
\tmp\Папка 1\Файл 1

>tar -cf tar.tar tmp 

>7z l tar.tar 

7-Zip 19.00 (x64) : Copyright (c) 1999-2018 Igor Pavlov : 2019-02-21

Scanning the drive for archives:
1 file, 10240 bytes (10 KiB)

Listing archive: tar.tar

--
Path = tar.tar
Type = tar
Physical Size = 10240
Headers Size = 10240
Code Page = UTF-8

   Date      Time    Attr         Size   Compressed  Name
------------------- ----- ------------ ------------  ------------------------
2020-10-24 22:11:02 D....            0            0  tmp
2020-10-24 22:10:56 D....            0            0  tmp\Folder 2
2020-10-24 22:10:56 .....            0            0  tmp\Folder 2\File 2
2020-10-24 22:10:56 D....            0            0  tmp\Папка 1
2020-10-24 22:10:56 .....            0            0  tmp\Папка 1\Файл 1
------------------- ----- ------------ ------------  ------------------------
2020-10-24 22:11:02                  0            0  2 files, 3 folders

>restic backup tmp 

Files:           0 new,     2 changed,     0 unmodified
Dirs:            0 new,     3 changed,     0 unmodified
Added to the repo: 1.184 KiB

processed 2 files, 0 B in 0:00
snapshot e82a9823 saved

>restic dump latest / >restic.tar 

>7z l restic.tar 

7-Zip 19.00 (x64) : Copyright (c) 1999-2018 Igor Pavlov : 2019-02-21

Scanning the drive for archives:
1 file, 5632 bytes (6 KiB)

Listing archive: restic.tar

--
Path = restic.tar
Type = tar
Physical Size = 5632
Headers Size = 4608
Code Page = UTF-8

   Date      Time    Attr         Size   Compressed  Name
------------------- ----- ------------ ------------  ------------------------
2020-10-24 22:11:02 D....            0            0  tmp
2020-10-24 22:10:56 D....            0            0  tmp\Folder 2
2020-10-24 22:10:56 .....            0            0  tmp\Folder 2\File 2
                    .....           26          512  tmp\ 1\PaxHeaders.0
2020-10-24 22:10:56 D....            0            0  tmp\ 1
                    .....           36          512  tmp\ 1\PaxHeaders.0\ 1
2020-10-24 22:10:56 .....            0            0  tmp\ 1\ 1
------------------- ----- ------------ ------------  ------------------------
2020-10-24 22:11:02                 62         1024  4 files, 3 folders
```

File: internal/dump/tar.go [hunk 1]
```diff
@@ -65,31 +66,59 @@ func tarTree(ctx context.Context, repo restic.Repository, rootNode *restic.Node,
...
 	if err != nil {
 		return err
 	}
 
+	fm := os.FileMode(node.Mode)
 	header := &tar.Header{
 		Name:       filepath.ToSlash(relPath),
 		Size:       int64(node.Size),
-		Mode:       int64(node.Mode),
+		Mode:       int64(fm.Perm()), // c_IS* constants are addde later
```
Comment (greatroar) [inline internal/dump/tar.go:88]: Typo: added

File: internal/dump/acl_test.go [hunk 2]
```diff
@@ -21,13 +21,27 @@ func Test_acl_decode(t *testing.T) {
...
+		},
 		{
 			name: "decode fail",
 			args: args{
 				xattr: []byte("abctest"),
 			},
 			want: "",
 		},
+		{
+			name: "decode fail",
```
Comment (greatroar) [inline internal/dump/acl_test.go:39]: Name copy-pasted from above. "empty"?

File: internal/dump/tar.go [hunk 3]
```diff
@@ -65,31 +66,59 @@ func tarTree(ctx context.Context, repo restic.Repository, rootNode *restic.Node,
...
+	c_ISVTX = 01000 // Save text (sticky bit)
+)
+
 func tarNode(ctx context.Context, tw *tar.Writer, node *restic.Node, repo restic.Repository) error {
 	relPath, err := filepath.Rel("/", node.Path)
 	if err != nil {
 		return err
 	}
 
+	fm := os.FileMode(node.Mode)
```
Comment (greatroar) [inline internal/dump/tar.go:84]: node.Mode is already of type os.FileMode. Also, mode is more obvious than fm.

File: internal/dump/tar.go (hunk 1 above)
Comment (MichaelEischer) [reply internal/dump/tar.go]: is fixed.

File: internal/dump/acl_test.go (hunk 2 above)
Comment (MichaelEischer) [reply internal/dump/acl_test.go]: I've changed the name to `decode empty fail`.

File: internal/dump/tar.go (hunk 3 above)
Comment (MichaelEischer) [reply internal/dump/tar.go]: is changed.

**Comment by MichaelEischer** [comment]
@DRON-666  The handling of non-ascii names belongs into a separate issue, I'm actually not sure whether it's at all possible to let the go library write tar headers using the ustar format (the only one supported by 7zip) for non-ascii filenames. Without this PR 7zip is not even able to read a tar file containing the dump of a directory at all.

**Comment by DRON-666** [comment]
>Without this PR 7zip is not even able to read a tar file containing the dump of a directory at all.

And it was better: can't open it in 7z, try open it in RAR. Now 7z open it successfully, but archive content totally broken.

File: changelog/unreleased/issue-2319 [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Bugfix: Correctly dump directories into tar files
+
+The dump command previously wrote directories in a tar file in a way which
+can cause compatibility problems. This caused for example 7zip on Windows
+to open tar files containing directories. In addition it was not possible
```
Comment (greatroar) [inline changelog/unreleased/issue-2319:5]: To NOT open tar files, I suppose? Also commas around ", for example," and comma after "In addition," (one more below).

File: changelog/unreleased/issue-2319 [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+to dump directories with extended attributes. These compatibility problems
+were fixed by properly setting the attributes of a directory.
```
Comment (greatroar) [inline changelog/unreleased/issue-2319:7]: Are now fixed: solution in present tense.

File: internal/dump/tar_test.go [hunk 6]
```diff
@@ -151,7 +164,17 @@ func checkTar(t *testing.T, testDir string, srcTar *bytes.Buffer) error {
 			if filepath.Base(hdr.Name) != filebase {
 				return fmt.Errorf("foldernames don't match got %v want %v", filepath.Base(hdr.Name), filebase)
 			}
-
+			if !strings.HasSuffix(hdr.Name, "/") {
+				return fmt.Errorf("foldernames must end with separator got %v", hdr.Name)
+			}
+		} else if hdr.Typeflag == tar.TypeSymlink {
```
Comment (greatroar) [inline internal/dump/tar_test.go:170]: Nitpick: is this the right moment to turn this into a switch statement?

File: changelog/unreleased/issue-2319 (hunk 4 above)
Comment (fd0) [reply changelog/unreleased/issue-2319:5]: I agree

File: changelog/unreleased/issue-2319 (hunk 5 above)
Comment (fd0) [reply changelog/unreleased/issue-2319:7]: Ok

File: internal/dump/tar_test.go (hunk 6 above)
Comment (fd0) [reply internal/dump/tar_test.go:170]: Yep, good point!

**Review by fd0 — CHANGES_REQUESTED** [review]

**Review by fd0 — APPROVED** [review]

**Pull request merged**